### ISSUE TITLE
Initiating 'Play' automatically scrolls the score to the top of the page. Vertical scrolling should be disabled.

### ISSUE THREAD

**Issue by Tantacrul** [musescore/MuseScore#10779]
**Describe the bug**
In this example (see video below), I have a large score with multiple instruments. Pressing play seems to vertically scroll the score back to the top of the page.

Obviously, this makes sense if the playback is occurring on a different system. For example, a solo piano score will require vertical scrolling if the user is zoomed in. However, when the system being played back is in view, there should be no vertical scrolling.

https://user-images.githubusercontent.com/21022311/157868158-f6ecf667-9cde-4d67-b709-500be858c814.mov

**Expected behaviour**
Where there is only one system on a page, the vertical position of the score should be retained during playback. This vertical position should be held (I.e. **not** changed) when playback moves to other pages containing only one system. 

Where playback moves to a page containing more than one system, the score should be repositioned according to present behaviour in MS4. 

See this document for further illustration of this: 
[playback-panning-v1.pdf](https://github.com/musescore/MuseScore/files/8581890/playback-panning-v1.pdf)

(N.B. I have tried to accurately paraphrase @MarcSabatella's descriptions below. @MarcSabatella please advise in the comments if any of this needs correcting. Also note that we are going for a lean fix here that addresses the present issue in situations with one system on a page only. Pages containing multiple systems should not be affected by this fix).

**Comment by Tantacrul** [comment]
I don't know why @wizofaus but I suspected you might have some good insights on this one. I imagine I've only uncovered one type of problem here.

**Comment by wizofaus** [comment]
Ok I have a very simple fix that basically says "don't zoom as long as the playback cursor is at least partly visible".
But that will mean even if only 1 pixel of it is visible, it won't zoom, as long as you're OK with that. @Tantacrul

**Comment by Tantacrul** [comment]
It sounds like a good step in the right direction that will solve this problem. Thanks!

**Comment by Tantacrul** [comment]
I've been encountering this again while writing an orchestral score in MS4. It is actually maddeningly frustrating, so I've promoted it up to a P1. Glad to see it's going to get triaged soon.

**Comment by DmitryArefiev** [comment]
@wizofaus I've tested the builds from #10799 and it looks not so good when the page has two systems:



https://user-images.githubusercontent.com/90187801/163822506-0aa07c07-d706-43a6-a719-770104d6a1b2.mp4

**Comment by MarcSabatella** [comment]
We tried hard to solve this for MuseScore 2 & 3, and make progress here and there, but it proved far more difficult to "do the right thing" in all the different cases than you could imagine.  Part of the problem was different expectations as to what the right thing is in some cases, like, continuing playback where you left off with nothing in particular selected, or with multiple systems per page but hide empty staves means they aren't all visible, etc.  Part of it was, the data structures needed to figure out how to position the page on the "right" place simply weren't available at the point in the code where the panning happened.  Continuous view simplifies this considerably and we generally recommend that when one wants to track a particular staff.

I suspect it's really only the special case of single-system pages where there is a fighting chance of getting this right always, so maybe that's where we focus.  A quick hack partial solution then could be, note at the start of playback if there if there is only a single system on the page, and if so, remember that Y position.  Then don't pan vertically unless you reach a page with more than one system, in which case do whatever, but as soon as you return to a page with only one system, restore the saved Y position.

**Comment by Tantacrul** [comment]
This is another one of those cases where all one has to do is open one of other major notation apps and observe that they already have decent rules for these things.

These things can easily be over designed but I'm sure there are quite a few relatively straightforward rules we can apply to get MS feeling better and not jerking the score to unexpected places as much as was true for v3.

**Comment by wizofaus** [comment]
> @wizofaus I've tested the builds from #10799 and it looks not so good when the page has two systems:
> 
>  bandicam.2022-04-18.17-13-10-240.mp4

Not sure I see the problem? This issue as I understood it was only about what happened when you *started* playback, not what happened as it continued to play and automatically scroll through the score.  I don't believe the change I made would have affected that.

**Comment by MarcSabatella** [comment]
Also it seems there should be logic that says, if the height of the system fits, do it.

As I understand it, the issue isn't just about starting playing but about following it - the idea that whatever staff or staves you are focused on (either because they were selected, or they were in view) when you start playback, that's what you'd like to keep in view if possible.  As mentioned, easier said than done when one considers all the different possibilities.  But the two easy to understand cases are, either the currently played system fits on screen and thus should, or else there is only one system per page and thus the originally viewed staves should remain in view where possible (or keep page physically at same vertical position).

But also, indeed, it would good to see what choice other programs have made in reach of these different situations.  My guess is they'll all behave similarly in the easy to understand, and we should do the same.  It's the other cases - cases with multiple systems per page, cases where at least some systems don't fit on screen, cases where the system height varies significantly, cases where the originally viewed system isn't even shown any more because it is hidden - those are the one where I expect to see no real consensus on the "right" answer, and where I'd say whatever we can come up with that doensn't break the easy to understand cases is probably acceptable.

**Comment by MarcSabatella** [comment]
BTW, I would note that MuseScore 3 does try to make those two optimizations work.  As long as your score remains one system per page, we keep the page at the same vertical position.  And if the current system fits on screen, we position it to do so.  But after that all bets are off.

MuseScore 4 appears to always reposition to show the top staff.  That means if a system fits on screen, position it at the top is the right thing to do - so in that case, it's as good as MuseScore 3.  But in the cases where there is only one system per [age (and it doesn't fit on screen), the jump to top is very bad.  It would be good to restore that particular optimization, so we're no worse than MuseScore 3 in this regard.

Bottom line, the change should be, if there is only one system on the page, keep the same Y offset, as was done in MuseScore 3.  This automatically does the right thing for large ensemble scores.

**Comment by wizofaus** [comment]
Logically you only want it to pan vertically if it's actually going to keep what you're currently looking at in view, plus reveal more of the same system.  E.g. if I'm specifically focused on the bottom staff of a system and start playback, I wouldn't mind if panned a bit to show more of the system, as long as it still shows that same bottom staff, which could, for instance, move to the bottom of the current view port.  If I'm focused on a staff right in the middle of the system (but the whole system isn't visible), I wouldn't really want it to pan at all.
So basically if the current staff is less than half way down the system, it's ok to auto pan upwards to bring more of the staves below into view (providing the current staff stays fully in view), and if it's more than half down the system, it's ok to auto pan downwards to bring more of the staves above into view.  If the system is already fully within the current viewport it shouldn't pan at all.

**Comment by Tantacrul** [comment]
@wizofaus do you need anything from us or is there a decision you'd like to be made?

Or are you happy with the PR and want it merged? If so, I'll give it a whirl!

**Comment by wizofaus** [comment]
My change introduces a regression as noted by Dmitry, so no, I wouldn't recommend merging as is.  I don't know what the right solution is to be honest.

**Comment by MarcSabatella** [comment]
My proposal for now is extremely simple: implement the same optimization we had in MuseScore 3 to the one-system-per-page case.

So, simply do not reposition the view vertically at all if there is only a single system on the page.  Only during playback - if the same function is used for other purposes, it should continue to function as usual.

For extra credit, make a note when playback *starts* if there is only one system on the page, and record the current vertical position.  Restore that vertical position any time you are on a page with only a single system.  This allows the expected result if some pages have multiple systems and some have only one - upon returning to a single-system page, you'd be back to the original vertical position.  MuseScore 3 didn't do that, so that's how to *improve* on existing behavior.

But for now at least, simply being no worse than MuseScore 3 should be easily achievable just by suppressing any vertical reposition for pages with only a single system, during playback only.

**Comment by Tantacrul** [comment]
@bkunda - when you're back from vacation, it might be good for you work out any kinks with Marc and we'll assign it to someone. I can't quite jump on it myself now.

**Comment by wizofaus** [comment]
I just tried it out in MS3 and I don't think it's great there, if you're zoomed some way in and have just been interacting with a particular staff towards the bottom of system on a 2-system page, starting playback bumps it out of view.

Really I would expect that it should only pan vertically if necessary to keep the staff that you've most recently interacted with in view.

**Comment by MarcSabatella** [comment]
Yes, as I said, with multiple systems per page that don't fit on screen, it is very difficult to always do the right thing.  That's the case that *doesn't* work in MU3 well, because it's hard.  But, two systems that *do* fit on screen is handled well in MU3 and in MU4 currently - just kee the entire system in view.  The previous proposed change, however, breaks this case.

MU3 does get the single-syste case right, and that's what this particular issue is about.  Right now, MU4 is very much worse than MU3 for this case, and the very simple fix I proposed addresses that.  And it does it without breaking the two-pages-that-fit-on-screen case.

With my proposed fix, it will still be true that two systems that *don't* fit on screen will present problems.  As I have said, it turns out to be more difficult than one might think to get this case "right", given all the possibilities.  So for now, I recommend not worrying about it.  Just fix the single-system case without breaking the two-systems-that-fit-on-screen case.  And my proposed fix does exactly that.  Keep everything exactly as is, but *don't* adjust vertical position on any page with one system - or better,  explicitly set the vertical position to what it was when playback started.

**Comment by bkunda** [comment]
I'm inclined to agree with @MarcSabatella in this case. 

For cases involving two systems or more, the fact that the entire system is kept in view works well. I think it's not unreasonable to let the user work out the appropriate zoom level in these cases to ensure that their entire system is visible during playback. A more sophisticated solution can wait until after MS4. 

So the real issue does seem to be single-system cases, where it does make sense to retain – across all pages – the vertical position (and zoom level) set at the time of playback commencement. 

@MarcSabatella I'd be happy to test whatever you or @wizofaus are able to implement to this end 🙂

**Comment by bkunda** [comment]
> My understanding is that @wizofaus has hit a snag with this.
> 
> I think we'll need to reassign it (@RomanPudashkin is on it now). @bkunda, to save poor Roman from having to read all of the comments above, can you be a point of contact for it? Let's just fix the bug raised here, as you say!

Absolutely. @RomanPudashkin we can chat today about this today if you like (I'm also happy just to provide a concise summary), and let me know if you have any specific questions as well :-)

**Comment by RomanPudashkin** [comment]
@bkunda please, update the description of the issue with all relevant behaviour requirements. It will be enough for me. You can also attach some document in which you describe the autoscroll behaviour

**Comment by bkunda** [comment]
> @bkunda please, update the description of the issue with all relevant behaviour requirements. It will be enough for me. You can also attach some document in which you describe the autoscroll behaviour

@RomanPudashkin I have added some "Expected behaviour" to the description above, along with a brief UX spec (PDF). Let me know if you have any questions, and thanks so much!

**Comment by RomanPudashkin** [comment]
I'm working on it

**Comment by boryskuzmenko** [comment]
🟢 FIXED
Tested by @Tantacrul
